### ISSUE TITLE
building with OpenMP complains local libpython installation is not PIC

### ISSUE THREAD

**Issue by beantowel** [llvm/llvm-project#64187]
I'm building `clang` and `openmp` using the `LLVM_ENABLE_PROJECTS` option:
```cmake
cmake -G Ninja -DLLVM_ENABLE_PROJECTS='clang;openmp' -DLLVM_ENABLE_PIC=ON -DCMAKE_BUILD_TYPE=Release ../llvm
cmake --build .
```
and the compiler complains:
```
FAILED: projects/openmp/libompd/gdb-plugin/python-module/ompd/ompdModule.so
: && ......<some falgs>..... /opt/local/lib/libpython3.9.a -ldl && :
.......ld: /opt/local/lib/libpython3.9.a(XXXX.o): relocation R_X86_64_32 against symbol XXX can not be used when making a shared object; recompile with -fPIC
```

The problem is, i have no permission to recompile and modify the system `libpython` build (and that's why i'm building clang form source code), do we have any options to remove the dependency on local libpython libraries?

**Comment by beantowel** [comment]
For those who got the same demand, i'm just skipping the `openmp/libompd/CMakeLists.txt`:
```cmake
........
# disable this block: if(LIBOMP_OMPD_SUPPORT)
if(0)
......
add_subdirectory(gdb-plugin)
......
endif()

```

**Comment by jhuber6** [comment]
Can you set `-DLIBOMP_OMPD_SUPPORT=OFF`? We should probably have an explicit option for this as I remember some previous issues related to the gdb plugin, @vigbalu.

**Comment by Shibodd** [comment]
I'm having the same exact issue, albeit building with  `-DLLVM_ENABLE_PROJECTS="clang" -DLLVM_ENABLE_RUNTIMES="openmp" `

CMake's FindPython3 prefers a static library version of libpython 3.5 when a dynamic library of version 3.8 is available.

Passing hints to FindPython3 by calling cmake with `-DPython3_USE_STATIC_LIBS="FALSE"` would resolve the issue in my case, but the flags are just used by the main CMake project which does not pass them to the runtimes subproject.

Any other solutions other than not building OMPD? In the meantime, i'll attempt to separate the two build steps.

**Comment by jhuber6** [comment]
> 

```diff
diff --git a/llvm/runtimes/CMakeLists.txt b/llvm/runtimes/CMakeLists.txt
index 75256c1326be..a49814efcd92 100644
--- a/llvm/runtimes/CMakeLists.txt
+++ b/llvm/runtimes/CMakeLists.txt
@@ -175,6 +175,8 @@ foreach(entry ${runtimes})
   list(APPEND prefixes ${canon_name})
   if (${canon_name} STREQUAL "OPENMP")
     list(APPEND prefixes "LIBOMP" "LIBOMPTARGET")
+    # The `openmp` runtime may require `-DPython` options to be recognized.
+    list(APPEND prefixes "Python")
   endif()
   # Many compiler-rt options start with SANITIZER_ and DARWIN_ rather than
   # COMPILER_RT_, so when compiler-rt is enabled, consider both.
```
Does this solve your problem?

**Comment by shiltian** [comment]
> ```diff
> ```diff
> diff --git a/llvm/runtimes/CMakeLists.txt b/llvm/runtimes/CMakeLists.txt
> index 75256c1326be..a49814efcd92 100644
> --- a/llvm/runtimes/CMakeLists.txt
> +++ b/llvm/runtimes/CMakeLists.txt
> @@ -175,6 +175,8 @@ foreach(entry ${runtimes})
>    list(APPEND prefixes ${canon_name})
>    if (${canon_name} STREQUAL "OPENMP")
>      list(APPEND prefixes "LIBOMP" "LIBOMPTARGET")
> +    # The `openmp` runtime may require `-DPython` options to be recognized.
> +    list(APPEND prefixes "Python")
>    endif()
>    # Many compiler-rt options start with SANITIZER_ and DARWIN_ rather than
>    # COMPILER_RT_, so when compiler-rt is enabled, consider both.
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> Does this solve your problem?
> ```

I think this looks quite confusing. We need to solve the issue in OpenMP to make it choose the right one instead of in LLVM.

**Comment by jhuber6** [comment]
> > ```diff
> > ```diff
> > diff --git a/llvm/runtimes/CMakeLists.txt b/llvm/runtimes/CMakeLists.txt
> > index 75256c1326be..a49814efcd92 100644
> > --- a/llvm/runtimes/CMakeLists.txt
> > +++ b/llvm/runtimes/CMakeLists.txt
> > @@ -175,6 +175,8 @@ foreach(entry ${runtimes})
> >    list(APPEND prefixes ${canon_name})
> >    if (${canon_name} STREQUAL "OPENMP")
> >      list(APPEND prefixes "LIBOMP" "LIBOMPTARGET")
> > +    # The `openmp` runtime may require `-DPython` options to be recognized.
> > +    list(APPEND prefixes "Python")
> >    endif()
> >    # Many compiler-rt options start with SANITIZER_ and DARWIN_ rather than
> >    # COMPILER_RT_, so when compiler-rt is enabled, consider both.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Does this solve your problem?
> 
> I think this looks quite confusing. We need to solve the issue in OpenMP to make it choose the right one instead of in LLVM.

I think it's a problem in general that we aren't respecting attempts to set which Python to use in CMake given that the plugins link directly against libpython.

**Comment by Shibodd** [comment]
Yes, this does solve the issue, with just a problem:

The variable Python3_COPYRIGHT is passed but formatted incorrectly resulting in ninja failing to build.

ninja output: `ninja: error: build.ninja:95512: expected '=', got identifier `

build.ninja, near line 95512:
```
build runtimes/runtimes-stamps/runtimes-configure | ${cmake_ninja_workdir}runtimes/runtimes-stamps/runtimes-configure: CUSTOM_COMMAND projects/runtimes/tmp/runtimes-cfgcmd.txt (...)
  COMMAND = cd /home/sbondi/workspace/llvm-work/build/runtimes/runtimes-bins && /usr/local/bin/cmake --no-warn-unused-cli -DCMAKE_C_COMPILER=  (...) -DCMAKE_CXX_COMPILER_WORKS=ON -DCMAKE_ASM_COMPILER_WORKS=ON -DHAVE_LLVM_LIT=ON -DLLVM_ENABLE_RUNTIMES=openmp -DLLVM_USE_LINKER=/home/sbondi/workspace/llvm-tools/bin/ld.lld "-DPython3_COPYRIGHT=Copyright (c) 2001-2021 Python Software Foundation.
All Rights Reserved. (<-- line 95512) 

Copyright (c) 2000 BeOpen.com.
All Rights Reserved.

Copyright (c) 1995-2001 Corporation for National Research Initiatives.
All Rights Reserved.

Copyright (c) 1991-1995 Stichting Mathematisch Centrum, Amsterdam.
All Rights Reserved." -DPython3_Compiler_FOUND=FALSE -DPython3_Development.Embed_FOUND=FALSE -DPython3_Development.Module_FOUND=FALSE (...) /home/sbondi/workspace/llvm-work/build/runtimes/runtimes-stamps/runtimes-configure
  DESC = Performing configure step for 'runtimes'
  pool = console
  restat = 1
```

Unsetting Python3_COPYRIGHT results in a successful compilation.
```diff
@@ -165,6 +165,8 @@ if(compiler_rt_path)
   endif()
 endif()
 
+unset(Python3_COPYRIGHT)
+
 # Create a list with the names of all the runtime projects in all uppercase and
 # with dashes turned to underscores. This gives us the CMake variable `prefixes`
 # for all variables that will apply to runtimes.
@@ -175,6 +177,7 @@ foreach(entry ${runtimes})
   list(APPEND prefixes ${canon_name})
   if (${canon_name} STREQUAL "OPENMP")
     list(APPEND prefixes "LIBOMP" "LIBOMPTARGET")
+    list(APPEND prefixes "Python")
   endif()
   # Many compiler-rt options start with SANITIZER_ and DARWIN_ rather than
   # COMPILER_RT_, so when compiler-rt is enabled, consider both.
```

**Comment by jhuber6** [comment]
Who it setting `-DPython_COPYRIGHT`? This should just be things set on the CMake invocation so it should be up to the user not to pass it if it causes problems I'd think.

**Comment by Shibodd** [comment]
It seems to be one of the result variables set by CMake's FindPython3.

https://github.com/Kitware/CMake/blob/72beefa91b7d6ea6eeffaf0977219e9e3fc4943d/Modules/FindPython/Support.cmake#L2211C10-L2224C18

**Comment by shiltian** [comment]
We might want to ask CMake community for this one? WDYT? @jhuber6

**Comment by jhuber6** [comment]
> We might want to ask CMake community for this one? WDYT? @jhuber6

Yeah, I don't think they should be setting a variable that expands so poorly when passed to another invocation.
